python/ray/train/session.py

```diff
@@ -301,6 +294,13 @@ def train_func():
         If no dataset is passed into Trainer, then return None.
     """
     session = get_session()
+    if session is None:
+        warnings.warn(
+            "`train.get_dataset_shard()` is meant to only be called "
+            "inside a training function that is executed by "
+            "`Trainer.run`."
+        )
+        return
```

Comment:
Would recommend encapsulating this in a helper function that takes 2 optional args: (1) return value, (2) warning message.

We can document this function so that its easy to know how to use properly (or not use in certain scenario) in future use-cases.

Can also centralize the log-once logic here.

Reply:
You could also implement this functionality using a decorator, although I'm not sure what the tradeoffs are for that approach verses a function.

Reply:
Good point, added a helper function for this. I think function over decorator should be fine for now out of simplicity